opentelemetry/tracing_interceptor.go

```diff
@@ -0,0 +1,202 @@
+	SpanStarter func(ctx context.Context, t trace.Tracer, spanName string, opts ...trace.SpanStartOption) trace.Span
+}
+
+type spanContextKey struct{}
+
+const defaultHeaderKey = "_tracer-data"
+
+type tracer struct{ options *TracerOptions }
+
+// NewTracer creates a tracer with the given options. Most callers should use
+// NewTracingInterceptor instead.
+func NewTracer(options TracerOptions) (interceptor.Tracer, error) {
+	if options.Tracer == nil {
+		options.Tracer = otel.GetTracerProvider().Tracer("temporal-sdk-go")
+	}
+	if options.TextMapPropagator == nil {
+		options.TextMapPropagator = DefaultTextMapPropagator
+	}
+	if options.SpanContextKey == nil {
+		options.SpanContextKey = spanContextKey{}
+	}
+	if options.HeaderKey == "" {
+		options.HeaderKey = defaultHeaderKey
+	}
+	if options.SpanStarter == nil {
+		options.SpanStarter = func(
+			ctx context.Context,
+			t trace.Tracer,
+			spanName string,
+			opts ...trace.SpanStartOption,
+		) trace.Span {
+			_, span := t.Start(ctx, spanName, opts...)
+			return span
+		}
+	}
+	return &tracer{&options}, nil
+}
+
+// NewTracingInterceptor creates an interceptor for setting on client options
+// that implements OpenTelemetry tracing for workflows.
+func NewTracingInterceptor(options TracerOptions) (interceptor.Interceptor, error) {
+	t, err := NewTracer(options)
+	if err != nil {
+		return nil, err
+	}
+	return interceptor.NewTracingInterceptor(t), nil
+}
+
+func (t *tracer) Options() interceptor.TracerOptions {
+	return interceptor.TracerOptions{
+		SpanContextKey: t.options.SpanContextKey,
+		HeaderKey:      t.options.HeaderKey,
+	}
+}
+
+func (t *tracer) UnmarshalSpan(m map[string]string) (interceptor.TracerSpanRef, error) {
+	ctx := trace.SpanContextFromContext(t.options.TextMapPropagator.Extract(context.Background(), textMapCarrier(m)))
+	if !ctx.IsValid() {
+		return nil, nil
+	}
+	return &tracerSpanRef{SpanContext: ctx}, nil
+}
+
+func (t *tracer) MarshalSpan(span interceptor.TracerSpan) (map[string]string, error) {
+	data := textMapCarrier{}
+	t.options.TextMapPropagator.Inject(trace.ContextWithSpan(context.Background(), span.(*tracerSpan).Span), data)
+	return map[string]string(data), nil
+}
+
+func (t *tracer) SpanFromContext(ctx context.Context) interceptor.TracerSpan {
+	span := trace.SpanFromContext(ctx)
+	if !span.SpanContext().IsValid() {
+		return nil
+	}
+	return &tracerSpan{Span: span}
+}
+
+func (t *tracer) ContextWithSpan(ctx context.Context, span interceptor.TracerSpan) context.Context {
+	return trace.ContextWithSpan(ctx, span.(*tracerSpan).Span)
+}
+
+func (t *tracer) StartSpan(opts *interceptor.TracerStartSpanOptions) (interceptor.TracerSpan, error) {
+	// Create context with parent
+	var parent trace.SpanContext
+	switch optParent := opts.Parent.(type) {
+	case nil:
+	case *tracerSpan:
+		parent = optParent.SpanContext()
+	case *tracerSpanRef:
+		parent = optParent.SpanContext
+	default:
+		panic(fmt.Sprintf("unrecognized parent type %T", optParent))
```

Comment:
This would be a compilation error if Go had generics. I had originally assumed the `StartSpan` call would only return runtime errors maybe for missing options or something. But I can return error here if we want.

Reply:
Ah, I see what you mean. Yeah, given this isn't an invariant that really *matters* to us a whole lot, I'm inclined to just return (or even just log) an error rather than exploding.

Reply:
Done